File: apps/agent/src/lib/gmail.ts
import { OAuth2Client } from "google-auth-library";
import { type Collection } from "mongodb";
import { getDb } from "./mongo.js";

/**
 * Gmail send integration — opt-in by env vars.
 *
 * Required env to enable real sending:
 *   GMAIL_OAUTH_CLIENT_ID
 *   GMAIL_OAUTH_CLIENT_SECRET
 *   GMAIL_OAUTH_REDIRECT_URI   (e.g. https://mnemos-agent.aryangorde.com/auth/google/callback)
 *
 * If any of these are missing, isGmailConfigured() returns false and the
 * approval flow falls back to simulated send (existing behaviour). This
 * means we can ship the code without breaking the demo.
 *
 * Per-user tokens are stored in mongo `gmail_tokens` keyed by userId.
 * For the single-user demo we hardcode userId = "alex".
 */

export const DEMO_USER_ID = "alex";

export const GMAIL_SCOPES = [
  "https://www.googleapis.com/auth/gmail.send",
  "https://www.googleapis.com/auth/userinfo.email",
  "openid",
];

export interface GmailTokenRecord {
  userId: string;
  accessToken: string;
  refreshToken: string;
  expiry: number; // ms epoch
  scope: string;
  email: string;
  updatedAt: Date;
}

export function isGmailConfigured(): boolean {
  return (
    !!process.env.GMAIL_OAUTH_CLIENT_ID &&
    !!process.env.GMAIL_OAUTH_CLIENT_SECRET &&
    !!process.env.GMAIL_OAUTH_REDIRECT_URI
  );
}

export function getOAuthClient(): OAuth2Client {
  if (!isGmailConfigured()) {
    throw new Error("Gmail OAuth env vars not configured");
  }
  return new OAuth2Client({
    clientId: process.env.GMAIL_OAUTH_CLIENT_ID,
    clientSecret: process.env.GMAIL_OAUTH_CLIENT_SECRET,
    redirectUri: process.env.GMAIL_OAUTH_REDIRECT_URI,
  });
}

export async function tokensCol(): Promise<Collection<GmailTokenRecord>> {
  const db = await getDb();
  return db.collection<GmailTokenRecord>("gmail_tokens");
}

export async function saveTokens(rec: Omit<GmailTokenRecord, "updatedAt">): Promise<void> {
  const col = await tokensCol();
  await col.updateOne(
    { userId: rec.userId },
    { $set: { ...rec, updatedAt: new Date() } },
    { upsert: true },
  );
}

export async function getTokens(userId: string): Promise<GmailTokenRecord | null> {
  const col = await tokensCol();
  return col.findOne({ userId });
}

/**
 * Returns a usable access token for the user — refreshes if expired.
 * Returns null if the user has no tokens, the OAuth env isn't configured,
 * or the refresh fails.
 */
export async function getAccessToken(userId: string): Promise<string | null> {
  if (!isGmailConfigured()) return null;
  const rec = await getTokens(userId);
  if (!rec) return null;

  const NOW = Date.now();
  // 60-second freshness buffer
  if (rec.expiry > NOW + 60_000) return rec.accessToken;

  // Refresh
  try {
    const client = getOAuthClient();
    client.setCredentials({ refresh_token: rec.refreshToken });
    const { credentials } = await client.refreshAccessToken();
    if (!credentials.access_token) return null;
    const newRec: Omit<GmailTokenRecord, "updatedAt"> = {
      userId: rec.userId,
      accessToken: credentials.access_token,
      refreshToken: credentials.refresh_token ?? rec.refreshToken,
      expiry: credentials.expiry_date ?? NOW + 3500 * 1000,
      scope: credentials.scope ?? rec.scope,
      email: rec.email,
    };
    await saveTokens(newRec);
    return newRec.accessToken;
  } catch {
    return null;
  }
}

/**
 * Send an email via the Gmail API. Returns the message id on success.
 * Throws on any send-side error so the caller can surface it.
 */
export async function sendGmail(
  userId: string,
  message: { to: string[]; cc?: string[]; subject: string; body: string },
): Promise<{ messageId: string; threadId?: string; sentAs: string }> {
  const accessToken = await getAccessToken(userId);
  if (!accessToken) throw new Error("gmail not connected — open /auth/google/start to authorize");

  const fromAddress = (await getTokens(userId))?.email ?? "me";

  // RFC 822 MIME
  const lines = [
    `To: ${message.to.join(", ")}`,
    ...(message.cc && message.cc.length > 0 ? [`Cc: ${message.cc.join(", ")}`] : []),
    `From: ${fromAddress}`,
    `Subject: ${message.subject}`,
    "MIME-Version: 1.0",
    "Content-Type: text/plain; charset=UTF-8",
    "",
    message.body,
  ];
  const raw = base64url(lines.join("\r\n"));

  const res = await fetch("https://gmail.googleapis.com/gmail/v1/users/me/messages/send", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ raw }),
  });
  if (!res.ok) {
    const detail = await res.text();
    throw new Error(`gmail.send ${res.status}: ${detail}`);
  }
  const data = (await res.json()) as { id: string; threadId?: string };
  return { messageId: data.id, ...(data.threadId ? { threadId: data.threadId } : {}), sentAs: fromAddress };
}

function base64url(s: string): string {
  return Buffer.from(s, "utf-8")
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

File: apps/agent-py/app/lib/gmail.py
"""Google OAuth + Gmail send — port of apps/agent/src/lib/gmail.ts.

Reads/writes the SAME `gmail_tokens` collection the TS backend uses, so an
existing connection works here with no re-auth. OAuth token exchange/refresh
done with raw HTTP (httpx) to avoid extra deps.
"""
from __future__ import annotations

import base64
import os
import time
from datetime import datetime, timezone

import httpx

from app.db.mongo import collection

DEMO_USER_ID = "alex"

GMAIL_SCOPES = [
    "https://www.googleapis.com/auth/gmail.send",
    "https://www.googleapis.com/auth/calendar.events",
    "https://www.googleapis.com/auth/userinfo.email",
    "openid",
]

_TOKEN_URL = "https://oauth2.googleapis.com/token"
_AUTH_URL = "https://accounts.google.com/o/oauth2/v2/auth"
_USERINFO_URL = "https://openidconnect.googleapis.com/v1/userinfo"


def is_gmail_configured() -> bool:
    return bool(
        os.environ.get("GMAIL_OAUTH_CLIENT_ID")
        and os.environ.get("GMAIL_OAUTH_CLIENT_SECRET")
        and os.environ.get("GMAIL_OAUTH_REDIRECT_URI")
    )


def _oauth_params() -> dict:
    return {
        "client_id": os.environ.get("GMAIL_OAUTH_CLIENT_ID", ""),
        "client_secret": os.environ.get("GMAIL_OAUTH_CLIENT_SECRET", ""),
        "redirect_uri": os.environ.get("GMAIL_OAUTH_REDIRECT_URI", ""),
    }


def tokens_col():
    return collection("gmail_tokens")


def auth_url() -> str:
    p = _oauth_params()
    from urllib.parse import urlencode
    q = urlencode({
        "client_id": p["client_id"],
        "redirect_uri": p["redirect_uri"],
        "response_type": "code",
        "scope": " ".join(GMAIL_SCOPES),
        "access_type": "offline",
        "prompt": "consent",
        "include_granted_scopes": "true",
    })
    return f"{_AUTH_URL}?{q}"


async def exchange_code(code: str) -> dict:
    p = _oauth_params()
    async with httpx.AsyncClient(timeout=20) as c:
        r = await c.post(_TOKEN_URL, data={
            "code": code, "client_id": p["client_id"], "client_secret": p["client_secret"],
            "redirect_uri": p["redirect_uri"], "grant_type": "authorization_code",
        })
        r.raise_for_status()
        return r.json()


async def fetch_email(access_token: str) -> str:
    async with httpx.AsyncClient(timeout=15) as c:
        r = await c.get(_USERINFO_URL, headers={"Authorization": f"Bearer {access_token}"})
        if r.status_code == 200:
            return r.json().get("email", "unknown@unknown")
    return "unknown@unknown"


async def save_tokens(rec: dict) -> None:
    rec = {**rec, "updatedAt": datetime.now(timezone.utc)}
    await tokens_col().update_one({"userId": rec["userId"]}, {"$set": rec}, upsert=True)


async def get_tokens(user_id: str) -> dict | None:
    return await tokens_col().find_one({"userId": user_id})


async def get_access_token(user_id: str) -> str | None:
    """Usable access token — refreshes if expired. None if unavailable."""
    if not is_gmail_configured():
        return None
    rec = await get_tokens(user_id)
    if not rec:
        return None
    now_ms = time.time() * 1000
    if rec.get("expiry", 0) > now_ms + 60_000:
        return rec.get("accessToken")
    # refresh
    p = _oauth_params()
    try:
        async with httpx.AsyncClient(timeout=20) as c:
            r = await c.post(_TOKEN_URL, data={
                "refresh_token": rec["refreshToken"], "client_id": p["client_id"],
                "client_secret": p["client_secret"], "grant_type": "refresh_token",
            })
            r.raise_for_status()
            tok = r.json()
        access = tok.get("access_token")
        if not access:
            return None
        new_rec = {
            "userId": rec["userId"],
            "accessToken": access,
            "refreshToken": tok.get("refresh_token", rec["refreshToken"]),
            "expiry": now_ms + tok.get("expires_in", 3500) * 1000,
            "scope": tok.get("scope", rec.get("scope", "")),
            "email": rec.get("email", ""),
        }
        await save_tokens(new_rec)
        return access
    except Exception:  # noqa: BLE001
        return None


def _b64url(s: str) -> str:
    return base64.urlsafe_b64encode(s.encode("utf-8")).decode("ascii").rstrip("=")


async def send_gmail(user_id: str, *, to: list[str], cc: list[str] | None,
                     subject: str, body: str) -> dict:
    access = await get_access_token(user_id)
    if not access:
        raise RuntimeError("gmail not connected — open /auth/google/start to authorize")
    rec = await get_tokens(user_id)
    from_addr = (rec or {}).get("email") or "me"
    lines = [f"To: {', '.join(to)}"]
    if cc:
        lines.append(f"Cc: {', '.join(cc)}")
    lines += [f"From: {from_addr}", f"Subject: {subject}",
              "MIME-Version: 1.0", "Content-Type: text/plain; charset=UTF-8", "", body]
    raw = _b64url("\r\n".join(lines))
    async with httpx.AsyncClient(timeout=30) as c:
        r = await c.post(
            "https://gmail.googleapis.com/gmail/v1/users/me/messages/send",
            headers={"Authorization": f"Bearer {access}", "Content-Type": "application/json"},
            json={"raw": raw},
        )
        if r.status_code >= 400:
            raise RuntimeError(f"gmail.send {r.status_code}: {r.text}")
        data = r.json()
    return {"messageId": data["id"], "threadId": data.get("threadId"), "sentAs": from_addr}
